Read mysql_hostgroup_attributes from proxysql configuration file

yashwantsahu20 opened the pull request:
Previously mysql_hostgroup_attributes was not read from configuration file.
Now mysql_hostgroup_attributes will be read from configuration file.
Issue for the same is #4552

yashwantsahu20 committed:
Previously mysql_hostgroup_attributes was not read from configuration file.
Now mysql_hostgroup_attributes will be read from configuration file.
Issue for the same is #4552

mirostauder commented:
Can one of the admins verify this patch?

yashwantsahu20 committed:
Read mysql_hostgroup_attributes from proxysql configuration file

It will insert fields only which are in configuration file.
If field data is invalid it will log a error.
It will save mysql_hostgroup_attributes into configuration file.

yashwantsahu20 committed:
1. Added tap test.
2. Saving to config file code correction

renecannao commented on test/tap/tests/mysql_hostgroup_attributes_config_file-t.cpp:
```diff
@@ -0,0 +1,123 @@
+	json j_servers_defaults { { "weight", 100 }, { "max_connections", 10000 }, { "use_ssl", 1}, };
+
+	// To run this test locally copy a ProxySQL config file into this path.
+	string save_config_query = {"SAVE CONFIG TO FILE /var/lib/jenkins//scripts/"
```
I don't think TAP tests should modify any config file. Config file should be immutable once the testing starts, for the whole duration of the CI run.

Thus a different approach could be:
* add `mysql_hostgroup_attributes` in the default configuration file . ( @mirostauder ? )
* the TAP test could verify that the configuration already loaded (at the beginning of the CI run, not at the beginning of the TAP test itself) is correct
* wipe the current table
* issue a new `LOAD MYSQL SERVERS FROM CONFIG`
* verify that the configuration is again what expected
* issue a `SAVE CONFIG TO FILE` on the TAP test working directory
* verify that the `mysql_hostgroup_attributes` section in the file is correct . The tap test itself could have hardcoded what to expect in the `mysql_hostgroup_attributes` section of the file

renecannao replied:
A possible enhancement (not strictly related to this PR) is that `LOAD MYSQL SERVERS FROM CONFIG` could accept a filename. And if this is the case, `Read_MySQL_Servers_from_configfile()` should accept a config file path.

yashwantsahu20 replied:
Thanks Rene. I will create a issue for the same.

renecannao commented on test/tap/tests/mysql_hostgroup_attributes_config_file-t.cpp:
```diff
@@ -0,0 +1,123 @@
+	string save_config_query = {"SAVE CONFIG TO FILE /var/lib/jenkins//scripts/"
+								"docker-mysql-proxysql/conf/proxysql/proxysql.cnf"};
+
+	const string INSERT_QUERY {
```
The test is quite complex in nature, but then it fails to "test enough" .
In this case it only insert one row , and only 3 columns out of 12 .
I think we should test all the 12 columns , and create more than just one hostgroup with (of course) different settings.

yashwantsahu20 replied:
Thanks fix it.

yashwantsahu20 committed:
it will read hostgroup_attributes from default config file and check correctness of it.
It will write hostgroup_attributes into config file and check correctness of it.

yashwantsahu20 committed:
Some small cleanup

yashwantsahu20 committed:
Changed the tables values based on actual docker config file.
